source/common/grpc/context_impl.cc

```diff
@@ -44,18 +60,19 @@ void ContextImpl::chargeStat(const Upstream::ClusterInfo& cluster, Protocol prot
   const Stats::StatName status_stat_name =
       (iter != stat_names_.status_names_.end()) ? iter->second : makeDynamicStatName(status_str);
   const Stats::SymbolTable::StoragePtr stat_name_storage =
-      symbol_table_.join({protocolStatName(protocol), request_names.service_, request_names.method_,
-                          status_stat_name});
+      request_names ? symbol_table_.join({protocolStatName(protocol), request_names->service_,
+                                          request_names->method_, status_stat_name})
+                    : symbol_table_.join({protocolStatName(protocol), status_stat_name});
 
   cluster.statsScope().counterFromStatName(Stats::StatName(stat_name_storage.get())).inc();
   chargeStat(cluster, protocol, request_names, (status_str == "0"));
 }
 
 void ContextImpl::chargeStat(const Upstream::ClusterInfo& cluster, Protocol protocol,
-                             const RequestNames& request_names, bool success) {
-  const Stats::SymbolTable::StoragePtr prefix_storage = symbol_table_.join(
-      {protocolStatName(protocol), request_names.service_, request_names.method_});
-  const Stats::StatName prefix(prefix_storage.get());
+                             const absl::optional<RequestNames>& request_names, bool success) {
+  auto prefix_and_storage = getPrefix(protocol, request_names);
+  Stats::StatName& prefix = prefix_and_storage.first;
```

Comment:
Think of `StatName` like a string_view; it's really a single pointer. No need to store by reference.